core/src/raw/oio/write/exact_buf_write.rs

```diff
@@ -47,8 +47,8 @@ impl<W: oio::Write> ExactBufWriter<W> {
 impl<W: oio::Write> oio::Write for ExactBufWriter<W> {
     async fn write(&mut self, mut bs: Buffer) -> Result<usize> {
         if self.buffer.len() >= self.buffer_size {
-            let written = self.inner.write(self.buffer.collect()).await?;
-            self.buffer.advance(written);
+            let buffer = std::mem::take(&mut self.buffer);
+            self.inner.write(buffer.collect()).await?;
```

Comment:
This is incorrect because `write` does not guarantee that all content will be successfully written. We need to `advance` the written size here and keep unwritten content.

Reply:
Then it's meaningless to make `collect` takes `mut self`. Please push a new commit on this patch or write a new PR.